tests/MongoDB.Driver.Tests/AsyncCursorTests.cs

```diff
@@ -78,17 +82,94 @@ public void KillCursor_should_actually_work()
+        public void Tailable_cursor_should_be_able_to_be_cancelled_from_a_different_thread_with_expected_result(
+            [Values(false, true)] bool async)
+        {
+            RequireServer.Check().Supports(Feature.TailableCursor);
+
+            string testCollectionName = "test";
+            string testDatabaseName = "test";
+
+            var eventCapturer = new EventCapturer().Capture<ConnectionReceivedMessageEvent>();
+            using (var disposedClient = DriverTestConfiguration.CreateDisposableClient(eventCapturer))
+            {
+                var client = disposedClient.Wrapped;
+                DropCollection(client, testDatabaseName, testCollectionName);
+                var database = client.GetDatabase(testDatabaseName);
+
+                var createCollectionOptions = new CreateCollectionOptions()
+                {
+                    Capped = true,
+                    MaxSize = 1000
+                };
+
+                database.CreateCollection(testCollectionName, createCollectionOptions);
+                var collection = database.GetCollection<BsonDocument>(testCollectionName);
+                collection.InsertOne(new BsonDocument());
+
+                var findOptions = new FindOptions<BsonDocument>()
+                {
+                    BatchSize = 1,
+                    CursorType = CursorType.TailableAwait
+                };
+
+                IAsyncCursor<BsonDocument> cursor;
+                if (async)
+                {
+                    cursor = collection
+                        .FindAsync(
+                            FilterDefinition<BsonDocument>.Empty,
+                            findOptions)
+                        .GetAwaiter()
+                        .GetResult();
+                }
+                else
+                {
+                    cursor = collection
+                        .FindSync(
+                            FilterDefinition<BsonDocument>.Empty,
+                            findOptions);
+                }
+
+                var cancellationTokenSource = new CancellationTokenSource();
+                var enumerator = cursor.ToEnumerable(cancellationTokenSource.Token).GetEnumerator();
+                enumerator.MoveNext();
+                eventCapturer.Clear();
+
+                var waitGetMoreAndCancelTask = eventCapturer
+                    .NotifyWhen(c => c.Any())
+                    .ContinueWith((task) =>
+                    {
+                        cancellationTokenSource.Cancel();
+                    });
+
+                var moveNextTask = Task.Run(() => enumerator.MoveNext());
+
+                var exception = Record.Exception(
+                    () =>
+                        Task
+                            .WhenAll(moveNextTask, waitGetMoreAndCancelTask)
+                            .GetAwaiter()
+                            .GetResult());
+                (exception is OperationCanceledException).Should().BeTrue();
```

Comment:
We can see two errors here: `OperationCanceledException` and `TaskCanceledException` (which is inherited from `OperationCanceledException`)

Reply:
The exception thrown from `enumerator.MoveNext()` should always be an `OperationCanceledException`.

The only reason you were seeing a `TaskCanceledException` was because of the way the test was written.